ykcapi/yk.h

```diff
@@ -48,4 +48,43 @@ YkLocation yk_location_new(void);
 // will occur.
 void yk_location_drop(YkLocation);
 
+void yk_stopgap(void *addr, uintptr_t size, uintptr_t retaddr, void *rsp);
+
+void __llvm_deoptimize(void* addr, uintptr_t size) {
+    // Push all registers to the stack before they can be clobbered, so that we
+    // can find their values after parsing in the stackmap. After that we store
+    // the value of RSP in a variable so we know where to look for the register
+    // values. The order in which we push the registers is equivalent to the
+    // Sys-V x86_64 ABI, which the stackmap format uses as well. However, since
+    // we are interested in the value of RSP in the previous stack frame, which
+    // has changed when we entered the __llvm_deoptimize function, we push RBP
+    // instead, which is equivalent. Note, that because of this there is no
+    // (easy) way to access RBP from the previous frame, so we just assume it's
+    // never referenced by the stackmap.
+    void *rsp;
+    __asm__ volatile (
+            "push %%rbp\n" // old RSP
+            "push %%rbp\n" // placeholder
+            "push %%rdi\n"
+            "push %%rsi\n"
+            "push %%rbx\n"
+            "push %%rcx\n"
+            "push %%rdx\n"
+            "push %%rax\n"
+            "movq %%rsp, %0\n": "=rm" (rsp));
+
+    // We can use this function's return address to lookup the correct record
+    // in the stackmap.
+    void *ret = __builtin_return_address(0);
+
+    // Forward stackmap address and size, and the return and rsp addresses to
+    // the function that parses the stackmap and initiates the stopgap
+    // interpreter.
+    yk_stopgap(addr, size, (uintptr_t) ret, rsp);
+
+    // Reverse the registers we've pushed onto the stack at the beginning of
```

Comment:
This doesn't look like it reverses anything to me, but perhaps I've misunderstood.

Reply:
I agree that it needs rewording.

Reply:
Fixed in 3e7cd91